larpandoracontent/LArReclustering/SimplePCAThreeDClusteringTool.cc

```diff
@@ -0,0 +1,120 @@
+    const CartesianVector axisDirection(eigenVecs.at(0).GetZ() > 0.f ? eigenVecs.at(0) : eigenVecs.at(0) * -1.f);
+
+    // Place intercept at hit with minimum projection
+    float minProjection(std::numeric_limits<float>::max());
+    for (const CaloHit *const pCaloHit3D : inputCaloHitList.get())
+        minProjection = std::min(minProjection, axisDirection.GetDotProduct(pCaloHit3D->GetPositionVector() - centroid));
+
+    const CartesianVector axisIntercept(centroid + (axisDirection * minProjection));
+
+    // Now define ortho directions
+    const CartesianVector seedDirection((axisDirection.GetX() < std::min(axisDirection.GetY(), axisDirection.GetZ())) ? CartesianVector(1.f, 0.f, 0.f) :
+        (axisDirection.GetY() < std::min(axisDirection.GetX(), axisDirection.GetZ())) ? CartesianVector(0.f, 1.f, 0.f) : CartesianVector(0.f, 0.f, 1.f));
+    const CartesianVector orthoDirection1(seedDirection.GetCrossProduct(axisDirection).GetUnitVector());
+
+    //Lists for hits that have a positive and negative projection on the secondary axis
+    CaloHitList posCaloHitList, negCaloHitList;
+
+    for (const CaloHit *const pCaloHit : inputCaloHitList.get())
+    {
+        const CartesianVector pCaloHitPosition = pCaloHit->GetPositionVector();
+
+        if((pCaloHitPosition-centroid).GetDotProduct(centroid+orthoDirection1)<0)
+        {
+            negCaloHitList.push_back(pCaloHit);
+        }
+		else
+		{ 
+            posCaloHitList.push_back(pCaloHit);
+		}
+    }
+
+    newCaloHitListsVector.push_back(posCaloHitList);
+    newCaloHitListsVector.push_back(negCaloHitList);
+
+    //Now, run PCA independently on pos and neg hit lists, to refine split
+
+    //Get pos list PCA
+    CartesianVector centroidPos(0.f, 0.f, 0.f);
+    LArPcaHelper::EigenVectors eigenVecsPos;
+    LArPcaHelper::EigenValues eigenValuesPos(0.f, 0.f, 0.f);
+    LArPcaHelper::RunPca(posCaloHitList, centroidPos, eigenValuesPos, eigenVecsPos);
+
+    // By convention, the primary axis has a positive z-component.
+    const CartesianVector axisDirectionPos(eigenVecsPos.at(0).GetZ() > 0.f ? eigenVecsPos.at(0) : eigenVecsPos.at(0) * -1.f);
+
+    //Get neg list PCA
+    CartesianVector centroidNeg(0.f, 0.f, 0.f);
+    LArPcaHelper::EigenVectors eigenVecsNeg;
+    LArPcaHelper::EigenValues eigenValuesNeg(0.f, 0.f, 0.f);
+    LArPcaHelper::RunPca(negCaloHitList, centroidNeg, eigenValuesNeg, eigenVecsNeg);
+
+    // By convention, the primary axis has a positive z-component.
+    const CartesianVector axisDirectionNeg(eigenVecsNeg.at(0).GetZ() > 0.f ? eigenVecsNeg.at(0) : eigenVecsNeg.at(0) * -1.f);
+
+    //Get intersection point of the two new principal axes
+    CartesianVector intersectionPoint(0.f, 0.f, 0.f);
+    float displacementPos(0.f),displacementNeg(0.f);
+    LArPointingClusterHelper::GetIntersection(centroidPos,axisDirectionPos,centroidNeg,axisDirectionNeg,intersectionPoint,displacementPos,displacementNeg);
```

Comment:
I think this needs to exist inside a try/catch.. sometimes an intersection won't exist and it looks in that case like LArPointingClusterHelper will throw an exception https://github.com/imawby/LArContent/blob/master/larpandoracontent/LArHelpers/LArPointingClusterHelper.cc